File: scenarios/basic/desktop/macos-appkit/LowContrastWarningText.swift
import Cocoa

class LowContrastWarningText: NSViewController {
    
    @IBOutlet weak var warningBanner: NSView!
    @IBOutlet weak var warningIcon: NSTextField!
    @IBOutlet weak var warningTitle: NSTextField!
    @IBOutlet weak var warningMessage: NSTextField!
    @IBOutlet weak var extendButton: NSButton!
    @IBOutlet weak var logoutButton: NSButton!
    @IBOutlet weak var flashCountLabel: NSTextField!
    
    @IBOutlet weak var dashboardTitle: NSTextField!
    @IBOutlet weak var dashboardSubtitle: NSTextField!
    @IBOutlet weak var timeRemainingLabel: NSTextField!
    
    @IBOutlet weak var quickActionsStackView: NSStackView!
    @IBOutlet weak var recentActivityStackView: NSStackView!
    
    private var flashCount = 0
    private var flashTimer: Timer?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        startFlashAnimation()
    }
    
    private func setupUI() {
        view.wantsLayer = true
        view.layer?.backgroundColor = NSColor.controlBackgroundColor.cgColor
        
        // Warning Banner - LOW CONTRAST WARNING TEXT
        warningBanner.wantsLayer = true
        warningBanner.layer?.backgroundColor = NSColor.systemRed.cgColor
        
        // LOW CONTRAST WARNING TEXT - Session timeout warning with insufficient color contrast
        warningIcon.stringValue = "⚠️"
        warningIcon.font = NSFont.systemFont(ofSize: 24)
        warningIcon.textColor = NSColor(red: 0.46, green: 0.46, blue: 0.46, alpha: 1.0) // #757575
        
        warningTitle.stringValue = "Session Timeout Warning"
        warningTitle.font = NSFont.boldSystemFont(ofSize: 16)
        warningTitle.textColor = NSColor(red: 0.46, green: 0.46, blue: 0.46, alpha: 1.0) // #757575
        
        warningMessage.stringValue = "Your session will expire in 5 minutes. Please save your work and extend your session to continue."
        warningMessage.font = NSFont.systemFont(ofSize: 14)
        warningMessage.textColor = NSColor(red: 0.74, green: 0.74, blue: 0.74, alpha: 1.0) // #BDBDBD
        
        // Action Buttons - LOW CONTRAST
        extendButton.title = "Extend"
        extendButton.bezelStyle = .rounded
        extendButton.target = self
        extendButton.action = #selector(extendSession)
        extendButton.font = NSFont.systemFont(ofSize: 14)
        extendButton.contentTintColor = NSColor(red: 0.56, green: 0.79, blue: 0.98, alpha: 1.0) // #90CAF9
        
        logoutButton.title = "Logout"
        logoutButton.bezelStyle = .rounded
        logoutButton.target = self
        logoutButton.action = #selector(logout)
        logoutButton.font = NSFont.systemFont(ofSize: 14)
        logoutButton.contentTintColor = NSColor(red: 0.94, green: 0.6, blue: 0.6, alpha: 1.0) // #EF9A9A
        
        flashCountLabel.stringValue = "Flashes: 0"
        flashCountLabel.font = NSFont.systemFont(ofSize: 12)
        flashCountLabel.textColor = .white
        flashCountLabel.backgroundColor = NSColor.black.withAlphaComponent(0.2)
        flashCountLabel.wantsLayer = true
        flashCountLabel.layer?.cornerRadius = 6
        
        // Dashboard Content
        dashboardTitle.stringValue = "Welcome back!"
        dashboardTitle.font = NSFont.systemFont(ofSize: 24, weight: .bold)
        
        dashboardSubtitle.stringValue = "Your session is active and secure"
        dashboardSubtitle.font = NSFont.systemFont(ofSize: 18)
        dashboardSubtitle.textColor = .secondaryLabelColor
        
        timeRemainingLabel.stringValue = "⏰ 5 minutes remaining"
        timeRemainingLabel.font = NSFont.systemFont(ofSize: 14)
        timeRemainingLabel.textColor = .secondaryLabelColor
        
        setupQuickActions()
        setupRecentActivity()
    }
    
    private func setupQuickActions() {
        let actions = [
            ("📊", "View Reports", "Learn new skills"),
            ("👥", "Manage Users", "User management"),
            ("⚙️", "Settings", "App configuration"),
            ("❓", "Help & Support", "Get assistance")
        ]
        
        for (icon, title, subtitle) in actions {
            let actionView = createActionView(icon: icon, title: title, subtitle: subtitle)
            quickActionsStackView.addArrangedSubview(actionView)
        }
    }
    
    private func setupRecentActivity() {
        let activities = [
            ("🟢", "User logged in", "2 minutes ago"),
            ("📊", "Report generated", "15 minutes ago"),
            ("⚙️", "Settings updated", "1 hour ago")
        ]
        
        for (icon, title, time) in activities {
            let activityView = createActivityView(icon: icon, title: title, time: time)
            recentActivityStackView.addArrangedSubview(activityView)
        }
    }
    
    private func createActionView(icon: String, title: String, subtitle: String) -> NSView {
        let containerView = NSView()
        containerView.wantsLayer = true
        containerView.layer?.backgroundColor = NSColor.controlBackgroundColor.cgColor
        containerView.layer?.cornerRadius = 8
        containerView.layer?.shadowOpacity = 0.2
        containerView.layer?.shadowOffset = NSSize(width: 0, height: 2)
        containerView.layer?.shadowRadius = 4
        
        let iconLabel = NSTextField()
        iconLabel.stringValue = icon
        iconLabel.font = NSFont.systemFont(ofSize: 32)
        iconLabel.isEditable = false
        iconLabel.isBordered = false
        iconLabel.backgroundColor = .clear
        
        let titleLabel = NSTextField()
        titleLabel.stringValue = title
        titleLabel.font = NSFont.boldSystemFont(ofSize: 16)
        titleLabel.isEditable = false
        titleLabel.isBordered = false
        titleLabel.backgroundColor = .clear
        
        let subtitleLabel = NSTextField()
        subtitleLabel.stringValue = subtitle
        subtitleLabel.font = NSFont.systemFont(ofSize: 12)
        subtitleLabel.textColor = .secondaryLabelColor
        subtitleLabel.isEditable = false
        subtitleLabel.isBordered = false
        subtitleLabel.backgroundColor = .clear
        
        containerView.addSubview(iconLabel)
        containerView.addSubview(titleLabel)
        containerView.addSubview(subtitleLabel)
        
        iconLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        subtitleLabel.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            containerView.heightAnchor.constraint(equalToConstant: 120),
            
            iconLabel.centerXAnchor.constraint(equalTo: containerView.centerXAnchor),
            iconLabel.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 16),
            
            titleLabel.centerXAnchor.constraint(equalTo: containerView.centerXAnchor),
            titleLabel.topAnchor.constraint(equalTo: iconLabel.bottomAnchor, constant: 8),
            
            subtitleLabel.centerXAnchor.constraint(equalTo: containerView.centerXAnchor),
            subtitleLabel.topAnchor.constraint(equalTo: titleLabel.bottomAnchor, constant: 4),
            subtitleLabel.bottomAnchor.constraint(lessThanOrEqualTo: containerView.bottomAnchor, constant: -16)
        ])
        
        return containerView
    }
    
    private func createActivityView(icon: String, title: String, time: String) -> NSView {
        let containerView = NSView()
        
        let iconLabel = NSTextField()
        iconLabel.stringValue = icon
        iconLabel.font = NSFont.systemFont(ofSize: 16)
        iconLabel.isEditable = false
        iconLabel.isBordered = false
        iconLabel.backgroundColor = .clear
        
        let titleLabel = NSTextField()
        titleLabel.stringValue = title
        titleLabel.font = NSFont.boldSystemFont(ofSize: 14)
        titleLabel.isEditable = false
        titleLabel.isBordered = false
        titleLabel.backgroundColor = .clear
        
        let timeLabel = NSTextField()
        timeLabel.stringValue = time
        timeLabel.font = NSFont.systemFont(ofSize: 12)
        timeLabel.textColor = .secondaryLabelColor
        timeLabel.isEditable = false
        timeLabel.isBordered = false
        timeLabel.backgroundColor = .clear
        
        containerView.addSubview(iconLabel)
        containerView.addSubview(titleLabel)
        containerView.addSubview(timeLabel)
        
        iconLabel.translatesAutoresizingMaskIntoConstraints = false
        titleLabel.translatesAutoresizingMaskIntoConstraints = false
        timeLabel.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            containerView.heightAnchor.constraint(equalToConstant: 40),
            
            iconLabel.leadingAnchor.constraint(equalTo: containerView.leadingAnchor),
            iconLabel.centerYAnchor.constraint(equalTo: containerView.centerYAnchor),
            iconLabel.widthAnchor.constraint(equalToConstant: 24),
            
            titleLabel.leadingAnchor.constraint(equalTo: iconLabel.trailingAnchor, constant: 8),
            titleLabel.centerYAnchor.constraint(equalTo: containerView.centerYAnchor),
            
            timeLabel.leadingAnchor.constraint(equalTo: titleLabel.trailingAnchor, constant: 8),
            timeLabel.centerYAnchor.constraint(equalTo: containerView.centerYAnchor),
            timeLabel.trailingAnchor.constraint(lessThanOrEqualTo: containerView.trailingAnchor)
        ])
        
        return containerView
    }
    
    private func startFlashAnimation() {
        flashTimer = Timer.scheduledTimer(withTimeInterval: 0.2, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            self.flashCount += 1
            self.flashCountLabel.stringValue = "Flashes: \(self.flashCount)"
            
            // Toggle between red and orange
            let currentColor = self.warningBanner.layer?.backgroundColor
            let newColor = currentColor == NSColor.systemRed.cgColor ? NSColor.systemOrange.cgColor : NSColor.systemRed.cgColor
            self.warningBanner.layer?.backgroundColor = newColor
        }
    }
    
    @objc private func extendSession() {
        // Handle extend session
        flashTimer?.invalidate()
        warningBanner.isHidden = true
    }
    
    @objc private func logout() {
        // Handle logout
        flashTimer?.invalidate()
        // Navigate to login screen
    }
    
    deinit {
        flashTimer?.invalidate()
    }
}

// MARK: - Storyboard Support
extension LowContrastWarningText {
    static func instantiate() -> LowContrastWarningText {
        let storyboard = NSStoryboard(name: "Main", bundle: nil)
        return storyboard.instantiateController(withIdentifier: "LowContrastWarningText") as! LowContrastWarningText
    }
}

File: scenarios/basic/desktop/linux-qt/LowContrastWarningText.py
import sys
from PyQt5.QtWidgets import (QApplication, QMainWindow, QVBoxLayout, QHBoxLayout, 
                             QWidget, QLabel, QPushButton, QFrame, QGridLayout, QScrollArea)
from PyQt5.QtCore import Qt, QTimer
from PyQt5.QtGui import QFont, QPalette, QColor

class LowContrastWarningText(QMainWindow):
    def __init__(self):
        super().__init__()
        self.setWindowTitle("Dashboard")
        self.setGeometry(100, 100, 1000, 700)
        
        # Central widget
        central_widget = QWidget()
        self.setCentralWidget(central_widget)
        
        # Main layout
        main_layout = QVBoxLayout(central_widget)
        main_layout.setSpacing(0)
        
        # Warning banner - LOW CONTRAST WARNING TEXT
        self.warning_frame = QFrame()
        self.warning_frame.setFixedHeight(80)
        self.warning_frame.setStyleSheet("QFrame { background-color: #ff0000; }")
        
        warning_layout = QHBoxLayout(self.warning_frame)
        warning_layout.setSpacing(15)
        warning_layout.setContentsMargins(20, 10, 20, 10)
        
        # Warning icon
        warning_icon = QLabel("⚠️")
        warning_icon.setStyleSheet("font-size: 24px;")
        warning_layout.addWidget(warning_icon)
        
        # Warning content
        warning_content_layout = QVBoxLayout()
        
        # LOW CONTRAST WARNING TEXT - Session timeout warning with insufficient color contrast
        warning_title = QLabel("Session Timeout Warning")
        title_font = QFont()
        title_font.setPointSize(16)
        title_font.setBold(True)
        warning_title.setFont(title_font)
        warning_title.setStyleSheet("color: #757575;")  # LOW CONTRAST
        warning_content_layout.addWidget(warning_title)
        
        warning_message = QLabel("Your session will expire in 5 minutes. Please save your work and extend your session to continue.")
        warning_message.setWordWrap(True)
        warning_message.setStyleSheet("color: #BDBDBD;")  # LOW CONTRAST
        warning_content_layout.addWidget(warning_message)
        
        warning_layout.addLayout(warning_content_layout)
        
        # Action buttons - LOW CONTRAST
        button_layout = QHBoxLayout()
        button_layout.setSpacing(10)
        
        extend_button = QPushButton("Extend")
        extend_button.setStyleSheet("QPushButton { color: #90CAF9; background-color: transparent; border: 1px solid #90CAF9; }")
        button_layout.addWidget(extend_button)
        
        logout_button = QPushButton("Logout")
        logout_button.setStyleSheet("QPushButton { color: #EF9A9A; background-color: transparent; border: 1px solid #EF9A9A; }")
        button_layout.addWidget(logout_button)
        
        # Flash count label
        self.flash_count_label = QLabel("Flashes: 0")
        self.flash_count_label.setStyleSheet("color: white; font-weight: bold; background-color: rgba(0, 0, 0, 0.2); padding: 4px 8px; border-radius: 6px;")
        button_layout.addWidget(self.flash_count_label)
        
        warning_layout.addLayout(button_layout)
        
        main_layout.addWidget(self.warning_frame)
        
        # Main content
        scroll_area = QScrollArea()
        scroll_area.setWidgetResizable(True)
        scroll_area.setVerticalScrollBarPolicy(Qt.ScrollBarAsNeeded)
        
        content_widget = QWidget()
        content_layout = QVBoxLayout(content_widget)
        content_layout.setSpacing(20)
        content_layout.setContentsMargins(20, 20, 20, 20)
        
        # Dashboard title
        dashboard_title = QLabel("Welcome back!")
        title_font = QFont()
        title_font.setPointSize(24)
        title_font.setBold(True)
        dashboard_title.setFont(title_font)
        dashboard_title.setAlignment(Qt.AlignCenter)
        content_layout.addWidget(dashboard_title)
        
        dashboard_subtitle = QLabel("Your session is active and secure")
        subtitle_font = QFont()
        subtitle_font.setPointSize(16)
        dashboard_subtitle.setFont(subtitle_font)
        dashboard_subtitle.setAlignment(Qt.AlignCenter)
        content_layout.addWidget(dashboard_subtitle)
        
        time_remaining = QLabel("⏰ 5 minutes remaining")
        time_remaining.setAlignment(Qt.AlignCenter)
        content_layout.addWidget(time_remaining)
        
        # Quick actions
        quick_actions_label = QLabel("Quick Actions")
        actions_font = QFont()
        actions_font.setPointSize(18)
        actions_font.setBold(True)
        quick_actions_label.setFont(actions_font)
        content_layout.addWidget(quick_actions_label)
        
        quick_actions_grid = QGridLayout()
        quick_actions_grid.setSpacing(20)
        
        actions = [
            ("📊", "View Reports", "Learn new skills"),
            ("👥", "Manage Users", "User management"),
            ("⚙️", "Settings", "App configuration"),
            ("❓", "Help & Support", "Get assistance")
        ]
        
        for i, (icon, title, subtitle) in enumerate(actions):
            action_frame = self.create_action_card(icon, title, subtitle)
            quick_actions_grid.addWidget(action_frame, i // 2, i % 2)
        
        content_layout.addLayout(quick_actions_grid)
        
        # Recent activity
        recent_activity_label = QLabel("Recent Activity")
        recent_activity_label.setFont(actions_font)
        content_layout.addWidget(recent_activity_label)
        
        recent_activity_frame = QFrame()
        recent_activity_frame.setFrameStyle(QFrame.Box)
        recent_activity_frame.setStyleSheet("QFrame { background-color: #f8f9fa; }")
        
        activity_layout = QVBoxLayout(recent_activity_frame)
        activity_layout.setSpacing(10)
        activity_layout.setContentsMargins(15, 15, 15, 15)
        
        activities = [
            ("🟢", "User logged in", "2 minutes ago"),
            ("📊", "Report generated", "15 minutes ago"),
            ("⚙️", "Settings updated", "1 hour ago")
        ]
        
        for icon, title, time in activities:
            activity_row = QHBoxLayout()
            
            icon_label = QLabel(icon)
            activity_row.addWidget(icon_label)
            
            title_label = QLabel(title)
            title_font = QFont()
            title_font.setBold(True)
            title_label.setFont(title_font)
            activity_row.addWidget(title_label)
            
            time_label = QLabel(time)
            time_label.setAlignment(Qt.AlignRight)
            activity_row.addWidget(time_label)
            
            activity_layout.addLayout(activity_row)
        
        content_layout.addWidget(recent_activity_frame)
        
        scroll_area.setWidget(content_widget)
        main_layout.addWidget(scroll_area)
        
        # Start flash animation
        self.flash_count = 0
        self.flash_timer = QTimer()
        self.flash_timer.timeout.connect(self.update_flash_display)
        self.flash_timer.start(200)  # 5 times per second
        
    def create_action_card(self, icon, title, subtitle):
        frame = QFrame()
        frame.setFrameStyle(QFrame.Box)
        frame.setFixedSize(200, 120)
        frame.setStyleSheet("QFrame { background-color: white; border: 1px solid #ddd; }")
        
        layout = QVBoxLayout(frame)
        layout.setSpacing(8)
        layout.setContentsMargins(15, 15, 15, 15)
        
        icon_label = QLabel(icon)
        icon_label.setAlignment(Qt.AlignCenter)
        icon_label.setStyleSheet("font-size: 36px;")
        layout.addWidget(icon_label)
        
        title_label = QLabel(title)
        title_font = QFont()
        title_font.setBold(True)
        title_label.setFont(title_font)
        title_label.setAlignment(Qt.AlignCenter)
        layout.addWidget(title_label)
        
        subtitle_label = QLabel(subtitle)
        subtitle_label.setAlignment(Qt.AlignCenter)
        subtitle_label.setWordWrap(True)
        layout.addWidget(subtitle_label)
        
        return frame
    
    def update_flash_display(self):
        self.flash_count += 1
        self.flash_count_label.setText(f"Flashes: {self.flash_count}")
        
        # Toggle between red and orange
        if self.flash_count % 2 == 0:
            self.warning_frame.setStyleSheet("QFrame { background-color: #ff0000; }")
        else:
            self.warning_frame.setStyleSheet("QFrame { background-color: #ff8800; }")

def main():
    app = QApplication(sys.argv)
    window = LowContrastWarningText()
    window.show()
    sys.exit(app.exec_())

if __name__ == "__main__":
    main()
